File: AdvancedDataStructureAndGraph/BinaryTree/pathSum.cpp
#include <unordered_map>
using namespace std;


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };


class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        unordered_map<long long, int> cnt;
        cnt[0] = 1;  // 前缀和为0出现1次（空路径），用于统计从根开始的路径

        return dfs(root, 0LL, (long long)targetSum, cnt);
    }

private:
    int dfs(TreeNode* node, long long curSum, long long target,
            unordered_map<long long, int>& cnt) {
        if (!node) return 0;

        curSum += node->val;

        // 以当前节点为终点，路径和为 target 的条数
        int res = 0;
        auto it = cnt.find(curSum - target);
        if (it != cnt.end()) res += it->second;

        // 把当前前缀和加入哈希表
        cnt[curSum]++;

        // 继续往下搜索
        res += dfs(node->left, curSum, target, cnt);
        res += dfs(node->right, curSum, target, cnt);

        // 回溯：撤销当前节点的前缀和记录
        cnt[curSum]--;

        return res;
    }
};

/*

算法思路（前缀和 + 计数哈希）

从根到当前节点的一条路径上，设：
    curSum = 从根到当前节点的路径和（前缀和）
    若存在某个祖先前缀和 prevSum，使得  
    curSum - prevSum = targetSum
    那么从“prevSum 后一个节点”到当前节点这段路径和就是 targetSum。

所以我们用哈希表 cnt 记录：某个前缀和出现过多少次：
    初始 cnt[0] = 1（表示“空路径前缀和”为 0，方便从根开始的路径计数）
    DFS 到节点：
        更新 curSum += node->val
        本节点作为路径终点的贡献：cnt[curSum - target]
        把 curSum 加入哈希：cnt[curSum]++
        递归左右子树
        回溯：cnt[curSum]--（撤销当前节点对路径的影响）

时空复杂度
设节点数为 n，树高为 h：
    时间复杂度：O(n)
        每个节点只做常数次哈希操作。
    空间复杂度：O(h)（平均）
        递归栈是 O(h)，哈希表最多存从根到当前路径上的前缀和数量，也是 O(h)；最坏退化链表时 O(n)

Q: long long 和 int 范围各是多少
A: 
    int 范围：-2,147,483,648 到 2,147,483,647 （约 -2.1e9 到 2.1e9）
    long long 范围：-9,223,372,036,854,775,808 到 9,223,372,036,854,775,807 （约 -9.2e18 到 9.2e18）
Q: hash 表key值为什么用 long long

   
*/

File: AdvancedDataStructureAndGraph/BinaryTree/pathSum.py
"""
Docstring for AdvancedDataStructureAndGraph.BinaryTree.pathSum

没思路>_<
"""


from collections import defaultdict

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        # cnt[p]：在“当前从根到正在访问节点的路径上”，前缀和为 p 出现的次数
        cnt = defaultdict(int)
        cnt[0] = 1  # 空路径前缀和=0，用于统计从根开始的路径

        def dfs(node: Optional[TreeNode], curSum: int) -> int:
            if node is None:
                return 0

            # 1) 更新前缀和：加入当前节点值
            curSum += node.val

            # 2) 以当前节点为终点，满足 targetSum 的路径条数
            #    需要 prevSum = curSum - targetSum
            res = cnt[curSum - targetSum]

            # 3) 将当前前缀和加入哈希表（进入该节点）
            cnt[curSum] += 1

            # 4) 继续向下搜索（路径必须向下）
            res += dfs(node.left, curSum)
            res += dfs(node.right, curSum)

            # 5) 回溯：离开该节点，撤销当前前缀和的记录（避免左右子树互相影响）
            cnt[curSum] -= 1

            return res

        return dfs(root, 0)

"""
chatgpt思路：
算法思路（前缀和 + 计数）

从根走到当前节点的路径和记为 curSum（前缀和）。
如果某条向下路径（从某个祖先的下一个节点开始，到当前节点结束）和为 targetSum，则：

curSum - prevSum = targetSum ⇒ prevSum = curSum - targetSum

所以我们用哈希表 cnt 记录“当前根到当前节点路径上，各个前缀和出现的次数”。
遍历到一个节点时：
    更新 curSum
    把 cnt[curSum - targetSum] 加到答案（表示以当前节点为终点的合法路径数）
    cnt[curSum] += 1（把当前前缀和加入路径）
    递归左右子树
    回溯 cnt[curSum] -= 1（离开该节点，撤销影响）

    

时空复杂度

设节点数为 n，树高为 h：
    - 时间复杂度：O(n)
        每个节点只进行常数次哈希访问/更新。

    - 空间复杂度：O(h)（平均）
        递归栈深度为 h，哈希表中同时有效的前缀和也最多 h 个；最坏退化链表为 O(n）。

没搞懂
搞懂了！好一个前缀和

Q&A
cnt = defaultdict(int)
    创建一个字典，默认值是 0
    当你访问一个不存在的 key 时，它会自动返回 0，而不是报错。

为什么 cnt[0] = 1 ？
    这是为了处理从根节点开始的路径。如果路径和正好等于 targetSum，那么 curSum - targetSum = 0。
    因此，我们需要预先在 cnt 中记录前缀和为 0 出现过一次，以便正确统计从根节点开始的路径。






"""
